### PR TITLE
MGMT-7234: InfraEnv Controller migrate to v2

### PR DESCRIPTION
**Pull request by rollandf**
# Assisted Pull Request

## Description
InfraEnv controller uses V2 API to update InfraEnv in backend

Signed-off-by: Fred Rolland <frolland@redhat.com>

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @filanov 
/cc @yevgeny-shnaidman 
/cc @nmagnezi 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/infraenv_controller.go [hunk 1]
```diff
@@ -201,133 +171,237 @@ func (r *InfraEnvReconciler) processNMStateConfig(ctx context.Context, log logru
...
+		proxy := &models.Proxy{
+			HTTPProxy:  &infraEnv.Spec.Proxy.HTTPProxy,
+			HTTPSProxy: &infraEnv.Spec.Proxy.HTTPSProxy,
+			NoProxy:    &infraEnv.Spec.Proxy.NoProxy,
+		}
+		createParams.InfraenvCreateParams.Proxy = proxy
+	}
 
+	if len(infraEnv.Spec.AdditionalNTPSources) > 0 {
+		createParams.InfraenvCreateParams.AdditionalNtpSources = swag.String(strings.Join(infraEnv.Spec.AdditionalNTPSources[:], ","))
```
Comment (yevgeny-shnaidman) [inline internal/controller/controllers/infraenv_controller.go:339]: @avishayt do we really need NTP sources in InfraEnv

File: internal/controller/controllers/infraenv_controller.go [hunk 2]
```diff
@@ -83,82 +83,52 @@ func (r *InfraEnvReconciler) Reconcile(origCtx context.Context, req ctrl.Request
...
+		proxy := &models.Proxy{}
+		internalProxy := internalInfraEnv.Proxy
+		if internalProxy == nil {
+			internalProxy = &models.Proxy{}
 		}
-		if infraEnv.Spec.Proxy.HTTPProxy != "" && infraEnv.Spec.Proxy.HTTPProxy != cluster.HTTPProxy {
-			params.HTTPProxy = swag.String(infraEnv.Spec.Proxy.HTTPProxy)
-			log.Debugf("HTTPProxy changed from %s to %s", cluster.HTTPProxy, infraEnv.Spec.Proxy.HTTPProxy)
-			update = true
+		if infraEnv.Spec.Proxy.NoProxy != "" && infraEnv.Spec.Proxy.NoProxy != swag.StringValue(internalProxy.NoProxy) {
```
[the remark this answers is not in the document]
Comment (rollandf) [reply internal/controller/controllers/infraenv_controller.go:101]: This is why I added the handling at line 96-99.
Should be enough, no?

[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/controller/controllers/infraenv_controller.go:101]: yes, missed it, scratch

File: internal/controller/controllers/infraenv_controller.go [hunk 3]
```diff
@@ -201,133 +171,237 @@ func (r *InfraEnvReconciler) processNMStateConfig(ctx context.Context, log logru
...
-	if inventoryErr != nil {
-		return r.handleEnsureISOErrors(ctx, log, infraEnv, inventoryErr)
+	if err = r.Installer.DeregisterInfraEnvInternal(ctx, installer.DeregisterInfraEnvParams{
+		InfraEnvID: infraEnv.ID,
+	}); err != nil {
+		return buildReply(err)
 	}
-	// Image successfully generated. Reflect that in infraEnv obj and conditions
-	return r.updateEnsureISOSuccess(ctx, log, infraEnv, updatedCluster.ImageInfo)
+	//TODO delete agents?
```
[the remark this answers is not in the document]
Comment (filanov) [reply internal/controller/controllers/infraenv_controller.go:384]: If user try to delete the namespace then we can assume that agents will be deleted as well so this operation will succeed. 
Another scenario is that hub cluster can drop the managed for installed hosts, so the spoke cluster will manage those hosts from now on. 
The strange UX is probably around a case when infra env is deleted when it still have not installed host that are bound to a cluster. I'm not sure if the user will understand why infra env is still needed in this case.
We should probably document this part, in addition we can prevent the deletion with webhooks.
@avishayt what do you think?

File: internal/controller/controllers/infraenv_controller.go [hunk 4]
```diff
@@ -83,82 +83,52 @@ func (r *InfraEnvReconciler) Reconcile(origCtx context.Context, req ctrl.Request
...
+		proxy := &models.Proxy{}
+		internalProxy := internalInfraEnv.Proxy
+		if internalProxy == nil {
+			internalProxy = &models.Proxy{}
 		}
-		if infraEnv.Spec.Proxy.HTTPProxy != "" && infraEnv.Spec.Proxy.HTTPProxy != cluster.HTTPProxy {
-			params.HTTPProxy = swag.String(infraEnv.Spec.Proxy.HTTPProxy)
-			log.Debugf("HTTPProxy changed from %s to %s", cluster.HTTPProxy, infraEnv.Spec.Proxy.HTTPProxy)
-			update = true
+		if infraEnv.Spec.Proxy.NoProxy != "" && infraEnv.Spec.Proxy.NoProxy != swag.StringValue(internalProxy.NoProxy) {
```
[the remark this answers is not in the document]
Comment (rollandf) [reply internal/controller/controllers/infraenv_controller.go:100]: according to @yevgeny-shnaidman no need to compare, just pass all params to backend .I will simplify the code here

[the remark this answers is not in the document]
Comment (filanov) [reply internal/controller/controllers/infraenv_controller.go:100]: Need to make sure that we will not generate the iso all the time, at least until we have the stream features merged.

File: internal/controller/controllers/infraenv_controller.go [hunk 5]
```diff
@@ -201,133 +154,235 @@ func (r *InfraEnvReconciler) processNMStateConfig(ctx context.Context, log logru
...
+
+	updateParams := installer.UpdateInfraEnvParams{
+		InfraEnvID:           infraEnvInternal.ID,
+		InfraEnvUpdateParams: &models.InfraEnvUpdateParams{},
+	}
+
+	// Check for updates from user and update the infraenv updateParams
+	r.updateInfraEnv(ctx, log, infraEnv, infraEnvInternal, updateParams.InfraEnvUpdateParams)
+
+	if cluster != nil {
```
[the remark this answers is not in the document]
Comment (filanov) [reply internal/controller/controllers/infraenv_controller.go:258]: It's ok to keep this code, we are currently targeting specific use case, infraenv is created with or without cluster id, we don't support editing it. so in worst case we will open a bug.

[the remark this answers is not in the document]
Comment (filanov) [reply internal/controller/controllers/infraenv_controller.go:258]: @avishayt do we need to support this case? 
Just to know if we need to fix that later.

File: internal/controller/controllers/infraenv_controller.go [hunk 6]
```diff
@@ -201,133 +154,235 @@ func (r *InfraEnvReconciler) processNMStateConfig(ctx context.Context, log logru
...
+		if clusterID.String() != infraEnvInternal.ClusterID.String() {
+			updateParams.InfraEnvUpdateParams.ClusterID = cluster.ID
+			log.Debugf("Cluster ID changed from %s to %s", infraEnvInternal.ClusterID, cluster.ID)
 		}
-		return ctrl.Result{Requeue: Requeue}, nil
 	}
 
-	// Check for updates from user, compare spec and update the cluster
-	if err = r.updateClusterIfNeeded(ctx, log, infraEnv, cluster); err != nil {
+	pullSecret, err := getPullSecretData(ctx, r.Client, r.APIReader, infraEnv.Spec.PullSecretRef, key.Namespace)
```
[the remark this answers is not in the document]
Comment (rollandf) [reply internal/controller/controllers/infraenv_controller.go]: Done

File: internal/controller/controllers/infraenv_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rollandf) [reply internal/controller/controllers/infraenv_controller.go]: Removed for now

**Review by filanov — APPROVED** [review]

**Pull request merged**